### ISSUE TITLE
Folding Core: GPU Checkpoint Option

### ISSUE THREAD

**Issue by Wazzzzuzp** [FoldingAtHome/fah-client-bastet#247]
Hi Devs,

Any chance of getting GPU Checkpoint Option added to the cores? Either to turn it off or extend it to x amount of frames? 

With my 4090 folding time per frame is normally in the 9 - 30 secs range. Having the core stop to perform a check point that takes as long as a frame to complete seems waste full. 

Thanks
![image](https://github.com/FoldingAtHome/fah-client-bastet/assets/135710766/9fb4ee2c-e094-41f0-a408-7df5c2b56246)

**Comment by muziqaz** [comment]
This is set by project owners, not the client. 4090 or other ultra high end users are extreme minority, and it is not very helpful to the projects to cater to that minority :)
Reducing the frequency of checkpoints would hinder science progress, because those with slower cards would lose a lot of progress if they paused or switched off folding just before it writes a checkpoint.
remember, your 9 seconds TPF is equal to minutes or tens of minutes to other performance tier cards. Checkpoints are not time based, too. so 5% for you is 45s, for others it might be half an hour, and they start losing half an hour of science, we are going backwards

**Comment by Wazzzzuzp** [comment]
Hence the ticket so the user can set the value.

**Comment by muziqaz** [comment]
CPU checkpointing is different, and is very fine grained. Regardless of what you set for yourself, CPU will always restart very close to where you paused/stopped. OpenMM (GPU), I believe, does not have such functionality. Whatever project owner (not fah dev) sets it in their project, it is universal and cannot be changed per client basis, at least with current client functionality. I believe if this was possible, this would have been implemented in v7 long time ago.

**Comment by Wazzzzuzp** [comment]
Better course of action is to go add a feature request to OpenMM?

**Comment by muziqaz** [comment]
No, better course of action is to be patient ;) interesting to hear what Joe has to say, and I also asked the question to GPU fahcore devs. Will see what comes out of it ;)

**Comment by jcoffland** [comment]
I don't think there's a problem here.  The checkpointing is taking less than one second.  This is apparent from the log.

**Comment by muziqaz** [comment]
We had a little chat internally and OpenMM checkpoint frequency is set by researcher at the beginning of the project, and is written into core.xml file. Not sure it would be ideal to have an option for users to alter those settings

**Comment by jcoffland** [comment]
The benefit to cost, in terms of performance gained vs development effort, is not good.

**Comment by Wazzzzuzp** [comment]
Better course of action is to go add a feature request to OpenMM?

> I don't think there's a problem here. The checkpointing is taking less than one second. This is apparent from the log.

Not the best example as this machine is running FAHClient from RAM.

Thanks all for entertain the idea.

**Comment by muziqaz** [comment]
That would still need to be exposed within fahclient for you to do the change

**Comment by Wazzzzuzp** [comment]
I dont follow?


> That would still need to be exposed within fahclient for you to do the change

Sorry I dont follow

**Comment by muziqaz** [comment]
Fahclient is the UI for users to adjust the settings. If fahclient does not have an actual option to change checkpoint frequency, you won't be able to change it ;)

**Comment by arisu3** [comment]
One important question is whether or not the sanity checks that occur for each checkpoint are factored in to the researcher's choice of checkpoint duration. If the sanity checks are only necessary for writing a valid checkpoint, then it should be simple enough to write a PR for OpenMM that would enable checkpointing at arbitrary times, but if the sanity checks are also important to ensure consistency of the simulation, then it might not be a good idea to allow control of the checkpoint rate, unless the sanity checks can be decoupled from the checkpoints.

**Comment by muziqaz** [comment]
I don't think sanity checks have anything to do with checkpoint decisions.
This is dependant on each individual project owner, and they desire to dig deeper into project set up. Majority just leave the standard checkpoint frequency in place. Sometimes we suggest the change if WU is very long and we would lose a lot of progress if we pause a WU.
Also l, the more frequent the checkpoint the larger the upload file is, which can seriously impact researchers storage space

**Comment by arisu3** [comment]
I checked now to confirm. The checkpoint file is serialized into XML, compressed, and uploaded. It contains version, step count, integrator information, parameters, energies, and periodic bounding box vectors for the simulation. It also contains positions, velocities, and forces for each particle, with three entries per particle. The checkpoint frequency won't affect final upload size unless it's combined with a greater XTC/TRR snapshot frequency, but the core.xml file specifies those independently.

The serialized checkpoint file contains the same information as the state.xml file, and its size is independent of checkpoint frequency, which makes sense since your checkpointState.xml file will be turned into the next generation's state.xml file.

**Comment by ghost** [comment]
All that aside, given most users with such a powerful GPU (4090, 5090, etc.) tend to have a performant SSD as well, it's still unnecessary writing to flash storage - which, as a matter of fact, has limited writing lifetime.

Given this fact alone should be _well_ enough to rectify the required development effort to either enable decreasing the number of checkpoints or, better yet, implement an algorithm that adjusts the number of checkpoints in accordance with the time per frame & maybe additionally the ETA, too.

After all, it's our hardware we - the majority of contributors - "donate" to this project & FAH should do its best to extend hardware lifetimes, especially in light of the crazy flash/memory prices these days.
